[68862] Create work package to submit project initiation request

cbliard opened the pull request:
# Ticket

https://community.openproject.org/wp/68862

# What are you trying to accomplish?

After project initiation request, create work package that will hold the pdf with submission information so that it can be approved.

# What approach did you choose and why?

Use a service

Not finished yet

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
Unsubscribe events when unloading JobStatus class in dev mode

`OpenProject::JobStatus::EventListener.register!` is called in  a
`config.to_prepare` block, meaning it's called after each reload in
development mode and events are subscribed to multiple times.

This is for instance visible in logs with the same job-related message
being logged multiple times:

```
I, [2025-11-21T10:02:13.993101 #51789]  INFO -- : [ActiveJob] Enqueued Journals::CompletedJob (Job ID: 85e5a6cf-e394-4d44-87c7-177aa4535a65) to GoodJob(default) at 2025-11-21 09:07:13 UTC
I, [2025-11-21T10:02:13.993971 #51789]  INFO -- : [ActiveJob] ↳ app/workers/journals/completed_job.rb:39:in 'Journals::CompletedJob.schedule'
D, [2025-11-21T10:02:13.994114 #51789] DEBUG -- : [ActiveJob] Enqueuing background job #<Journals::CompletedJob:0x00000003d59def80 @arguments=[81, 2025-11-21 09:00:52.603688000 UTC +00:00, true], @job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @queue_name="default", @scheduled_at=2025-11-21 09:07:13.984857000 UTC +00:00, @priority=5, @executions=0, @exception_executions={}, @timezone="Etc/UTC", @successfully_enqueued=true, @provider_job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @_halted_callback_hook_called=nil>
D, [2025-11-21T10:02:13.994153 #51789] DEBUG -- : [ActiveJob] Enqueuing background job #<Journals::CompletedJob:0x00000003d59def80 @arguments=[81, 2025-11-21 09:00:52.603688000 UTC +00:00, true], @job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @queue_name="default", @scheduled_at=2025-11-21 09:07:13.984857000 UTC +00:00, @priority=5, @executions=0, @exception_executions={}, @timezone="Etc/UTC", @successfully_enqueued=true, @provider_job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @_halted_callback_hook_called=nil>
D, [2025-11-21T10:02:13.994178 #51789] DEBUG -- : [ActiveJob] Enqueuing background job #<Journals::CompletedJob:0x00000003d59def80 @arguments=[81, 2025-11-21 09:00:52.603688000 UTC +00:00, true], @job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @queue_name="default", @scheduled_at=2025-11-21 09:07:13.984857000 UTC +00:00, @priority=5, @executions=0, @exception_executions={}, @timezone="Etc/UTC", @successfully_enqueued=true, @provider_job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @_halted_callback_hook_called=nil>
D, [2025-11-21T10:02:13.994203 #51789] DEBUG -- : [ActiveJob] Enqueuing background job #<Journals::CompletedJob:0x00000003d59def80 @arguments=[81, 2025-11-21 09:00:52.603688000 UTC +00:00, true], @job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @queue_name="default", @scheduled_at=2025-11-21 09:07:13.984857000 UTC +00:00, @priority=5, @executions=0, @exception_executions={}, @timezone="Etc/UTC", @successfully_enqueued=true, @provider_job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @_halted_callback_hook_called=nil>
D, [2025-11-21T10:02:13.994229 #51789] DEBUG -- : [ActiveJob] Enqueuing background job #<Journals::CompletedJob:0x00000003d59def80 @arguments=[81, 2025-11-21 09:00:52.603688000 UTC +00:00, true], @job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @queue_name="default", @scheduled_at=2025-11-21 09:07:13.984857000 UTC +00:00, @priority=5, @executions=0, @exception_executions={}, @timezone="Etc/UTC", @successfully_enqueued=true, @provider_job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @_halted_callback_hook_called=nil>
D, [2025-11-21T10:02:13.994254 #51789] DEBUG -- : [ActiveJob] Enqueuing background job #<Journals::CompletedJob:0x00000003d59def80 @arguments=[81, 2025-11-21 09:00:52.603688000 UTC +00:00, true], @job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @queue_name="default", @scheduled_at=2025-11-21 09:07:13.984857000 UTC +00:00, @priority=5, @executions=0, @exception_executions={}, @timezone="Etc/UTC", @successfully_enqueued=true, @provider_job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @_halted_callback_hook_called=nil>
D, [2025-11-21T10:02:13.994278 #51789] DEBUG -- : [ActiveJob] Enqueuing background job #<Journals::CompletedJob:0x00000003d59def80 @arguments=[81, 2025-11-21 09:00:52.603688000 UTC +00:00, true], @job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @queue_name="default", @scheduled_at=2025-11-21 09:07:13.984857000 UTC +00:00, @priority=5, @executions=0, @exception_executions={}, @timezone="Etc/UTC", @successfully_enqueued=true, @provider_job_id="85e5a6cf-e394-4d44-87c7-177aa4535a65", @_halted_callback_hook_called=nil>
...
```

Fix it by unsubscribing them all when the class is unloaded.

cbliard committed:
Create artifact work package on project initiation request submission

https://community.openproject.org/wp/68862

A service takes care of creating the artifact work package.

cbliard committed:
Ensure no missing translations in contract tests with `:check_errors_i18n` metadata

When `:check_errors_i18n` is added to example metadata,
`expect_contract_invalid` uses `full_messages` to check for missing
translations in tests.

When an error is translated, it needs two translations to exist:
- the attribute, with translation being searched in following paths:
  - activerecord.attributes.[model].[attribute_name]
  - attributes.[attribute_name]
  - activerecord.models.[attribute_name]
- the error, with translation being searched in following paths:
  - activerecord.errors.models.[model_name].attributes.[attribute_name].[error]
  - activerecord.errors.models.[model_name].[error]
  - activerecord.errors.messages.[error]
  - errors.attributes.[attribute_name].[error]
  - errors.messages.[error]

It can't be enabled globally because some keys do not exist in some
tests scenarios.

oliverguenther committed:
Create PDF as attachment

oliverguenther committed:
Redirect to wizard if enabled

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/68862-create-work-package-to-submit-project-initiation-request

cbliard committed:
Check :add_work_package permission for artifact work package creation

cbliard committed:
Ensure only one error is shown if PIR is disabled

When Project Initiation Request is disabled, no need to display errors about the status being not set (for instance).

cbliard committed:
Fix test by having a project with valid PIR info

cbliard committed:
Fix rubocop warnings about ABC metrics size

cbliard committed:
Assign artifact work package to the designated user

In a project initation request, the created artifact work package is
assigned to the user pointed by a user custom field. The user custom
field to get this user from is given by
`project.project_creation_wizard_assignee_custom_field_id`.

cbliard committed:
Test type, status, and attachment of PIR work package

cbliard committed:
Set artifact work package subject to the name configured in settings

cbliard committed:
Merge branch 'dev' into feature/68862-create-work-package-to-submit-project-initiation-request

cbliard committed:
Fix tests related to artifact work package creation

A project settings attribute name was renamed from `name_artefact_name`
to `project_creation_wizard_artifact_name`.

cbliard committed:
Comment and mention the assignee in the artifact work package

cbliard committed:
Use default description for artifact work package

This way, if it has a template content, it will be used.

oliverguenther commented on app/controllers/projects/creation_wizard_controller.rb:
```diff
@@ -54,14 +54,23 @@ def update # rubocop:disable Metrics/AbcSize
+          flash[:error] = service_call2.errors.full_messages
+          render :show,
+                 locals: { menu_name: :none },
+                 status: :unprocessable_entity
```
```suggestion
        creation_call = Projects::CreateArtifactWorkPackageService.new(user: current_user, model: @project).call
        if creation_call.success?
          redirect_to project_work_packages_path(@project, @project.project_creation_wizard_artifact_work_package_id),
                      notice: I18n.t("projects.wizard.success")
        else
          flash[:error] = creation_call.errors.full_messages
          render :show,
                 locals: { menu_name: :none },
                 status: :unprocessable_entity
```

I'd prefer an explicit name over just `service_call2`

cbliard replied:
ah yes, that's a part I wrote quickly initially to get things working. It was not meant to stay like this.

Good that you spotted it.

oliverguenther committed:
Use journal_notes directly

cbliard committed:
Use better name for call result

Co-authored-by: Oliver Günther <o.guenther@openproject.com>

cbliard committed:
Remove unneeded permission check for comment edition

Now that the work package is created directly with the right comment
through `journal_notes`, less permissions are needed.
